New class for triangle

lbartoletti opened the pull request:
Hello,

I propose this new class which comes in part from my old plugin (CADDigitize). Almost all the code comes from this one.
I think, that this class is a first step for the OGC specification.

There are still tests and additions to make, but I prefer to propose it to you now to have your return.

Regards

lbartoletti committed:
init QgsTriangle

lbartoletti committed:
Merge branch 'master' of https://github.com/QGIS/QGIS into qgstriangle

lbartoletti committed:
Add fromwkb/wkt with tests. Add some notes and todo

lbartoletti committed:
Add some methods for triangle and geometryUtils for this methods.

lbartoletti committed:
Merge branch 'master' of https://github.com/QGIS/QGIS into qgstriangle

lbartoletti committed:
Add: orthocenter, bisectors, medians, medial, circumscribed (center,
radius), inscribed (center, radius)

lbartoletti committed:
Merge branch 'master' of https://github.com/QGIS/QGIS into qgstriangle

rouault commented on src/core/geometry/qgswkbtypes.h:
```diff
@@ -114,9 +117,11 @@ class CORE_EXPORT QgsWkbTypes
       Point25D = 0x80000001,
       LineString25D,
       Polygon25D,
+      Triangle25D,
```
Adding Triangle25D here will break the values of later enumeration items. xxx25D are for legacy stuff. No need for that for newer geometry types

nyalldawson commented on src/core/geometry/qgsgeometryutils.h:
```diff
@@ -291,6 +291,31 @@ class CORE_EXPORT QgsGeometryUtils
      */
     static QgsPointV2 midpoint( const QgsPointV2& pt1, const QgsPointV2& pt2 );
 
+    /** Return the slope of a line defined by points pt1 and pt2.
```
Small style change, but we're moving away from explicit listing of all params in cases like this (where the meaning has already been described). It's an nice touch to change "defined by points P1 and p2" to "defined by points \a p1 and \a p2" so that doxygen formats these as b arguments.

nyalldawson commented on src/core/geometry/qgsgeometryutils.h:
```diff
@@ -291,6 +291,31 @@ class CORE_EXPORT QgsGeometryUtils
+    /** Return the slope of a line defined by points pt1 and pt2.
+     * @param pt1 first point.
+     * @param pt2 second point.
+     * @return The slope of this linear entity, or infinity if vertical
```
Probably worth mentioning that the slope returned is a gradient, and not an angle

nyalldawson replied:
Actually - maybe better to rename this method as "gradient" for clarity.

nyalldawson commented on src/core/geometry/qgsgeometryutils.h:
```diff
@@ -291,6 +291,31 @@ class CORE_EXPORT QgsGeometryUtils
+     * @return The coefficients of this linear entity
+     * @note added in QGIS 3.0
+     */
+    static QVector<double> coefficients( const QgsPointV2& pt1, const QgsPointV2& pt2 );
```
Instead of returning a vector I think a/b/c should be arguments by reference

lbartoletti replied:
OK. Do you want the same correction for:
QVector<double> lengths() const;
QVector<double> angles() const;
QVector<QgsLineString *> altitudes( ) const;
QVector<QgsLineString *> medians( ) const; 
QVector<QgsLineString *> bisectors( double lengthTolerance = 0.0001 ) const;

nyalldawson replied:
No, I think those are fine to return as lists since the order isn't important

nyalldawson commented on python/core/geometry/qgsgeometryutils.sip:
```diff
@@ -67,4 +67,10 @@ class QgsGeometryUtils
+
+    static QVector<double> coefficients(const QgsPointV2 &pt1, const QgsPointV2 &pt2);
+
+    static QgsLineString* perpendicularSegment( const QgsPointV2& p, const QgsPointV2& s1, const QgsPointV2& s2 );
```
Needs a "/Factory/" annotation

nyalldawson commented on src/core/geometry/qgsgeometryutils.cpp:
```diff
@@ -834,6 +834,82 @@ QgsPointV2 QgsGeometryUtils::midpoint( const QgsPointV2 &pt1, const QgsPointV2 &
+
+QgsLineString* QgsGeometryUtils::perpendicularSegment( const QgsPointV2 &p, const QgsPointV2 &s1, const QgsPointV2 &s2 )
+{
+  QgsLineString *line = new QgsLineString;
```
I'd like to see us move toward the more modern c++11 approach of using a unique_ptr here, which is only released when the object is returned

nyalldawson commented on src/core/geometry/qgstriangle.cpp:
```diff
@@ -0,0 +1,421 @@
+
+  mWkbType = parts.first;
+
+  QString defaultChildWkbType = QStringLiteral( "LineString%1%2" ).arg( is3D() ? "Z" : "", isMeasure() ? "M" : "" );
```
Use QString() instead of "" here

nyalldawson reviewed (commented):
Looking great so far!

lbartoletti committed:
Remove Triangle25D (change request)

lbartoletti committed:
Code changed according to Nyall's requests

nyalldawson commented:
Looking great to me! Do you have a to-do list for remaining changes before this is merge ready?

lbartoletti commented:
My TODO:
- Add python binding
- Add triangle for QgsExpression
- Write full (best) doc
- Add some tests
- Check constructor with 3 points (no double point, colinearity). Maybe other constructors (QgsPoint, QPointF)

I'll do it at the weekend and you can merge it if it's ok.

Future TODO:
- export circumscribed and inscribed circle to QgsCircle (new class. I will work on it)

lbartoletti committed:
- Python binding added
- Tests colinearity and double points for constructor with 3 points

lbartoletti committed:
Add triangle (make_triangle) into qgsexpression.cpp

lbartoletti committed:
- altitudes, medians and bisectors methods are not available in Python
bindings (crash QGis)
- documentation writed
- tests added for "deleted" methods

nyalldawson commented:
Do you still get a crash if you change altitudes, medians and bisectors to return QVector< QgsLineString > (ie not pointers)? I think this might be a safer choice in any case.

nyalldawson commented on src/core/qgsexpression.cpp:
```diff
@@ -3991,6 +4024,7 @@ const QList<QgsExpression::Function*>& QgsExpression::Functions()
     << new StaticFunction( QStringLiteral( "make_point_m" ), 3, fcnMakePointM, QStringLiteral( "GeometryGroup" ) )
     << new StaticFunction( QStringLiteral( "make_line" ), -1, fcnMakeLine, QStringLiteral( "GeometryGroup" ) )
     << new StaticFunction( QStringLiteral( "make_polygon" ), -1, fcnMakePolygon, QStringLiteral( "GeometryGroup" ) )
+    << new StaticFunction( QStringLiteral( "make_triangle" ), -1, fcnMakeTriangle, QStringLiteral( "GeometryGroup" ) )
```
Can you change this to use the new style which takes a ParameterList? That will correctly handle detection of invalid arguments.

nyalldawson replied:
There's one part missing from this commit - you also need to add function help in https://github.com/qgis/QGIS/tree/master/resources/function_help/json

nyalldawson commented on src/core/qgsexpression.cpp:
```diff
@@ -2187,6 +2188,38 @@ static QVariant fcnMakePolygon( const QVariantList& values, const QgsExpressionC
+  {
+    QgsGeometry geom = getGeometry( value, parent );
+    if ( geom.isNull() )
+      continue;
```
What happens in the case that any of these continues are hit? Won't you be setting the exterior ring to a linestring with < 3 vertices?

I think it's better to flag an evaluation error if any of these conditions are hit.

nyalldawson commented on src/core/qgsexpression.cpp:
```diff
@@ -2187,6 +2188,38 @@ static QVariant fcnMakePolygon( const QVariantList& values, const QgsExpressionC
+    return QVariant();
+  }
+
+  QgsTriangle *tr = new QgsTriangle();
```
switch these two to unique_ptrs for memory safety

nyalldawson commented on tests/src/core/testqgsexpression.cpp:
```diff
@@ -765,6 +765,8 @@ class TestQgsExpression: public QObject
       QTest::newRow( "make_polygon rings" ) << "geom_to_wkt(make_polygon(geom_from_wkt('LINESTRING( 0 0, 0 1, 1 1, 1 0, 0 0 )'),geom_from_wkt('LINESTRING( 0.1 0.1, 0.1 0.2, 0.2 0.2, 0.2 0.1, 0.1 0.1 )'),geom_from_wkt('LINESTRING( 0.8 0.8, 0.8 0.9, 0.9 0.9, 0.9 0.8, 0.8 0.8 )')))" << false
       << QVariant( "Polygon ((0 0, 0 1, 1 1, 1 0, 0 0),(0.1 0.1, 0.1 0.2, 0.2 0.2, 0.2 0.1, 0.1 0.1),(0.8 0.8, 0.8 0.9, 0.9 0.9, 0.9 0.8, 0.8 0.8))" );
+      QTest::newRow( "make_triangle bad" ) << "make_triangle(make_point(2,4))" << false << QVariant();
+      QTest::newRow( "make_triangle" ) << "geom_to_wkt(make_triangle(make_point(0,0), make_point(5,5), make_point(0,10)))" << false << QVariant( "Triangle ((0 0, 5 5, 0 10, 0 0))" );
```
Can you also add tests which hit the three conditions checked in the loop? Ie null geometry/non-point geometries.

lbartoletti committed:
Change methods altitudes, medians, bisectors from QVector<QgsLineString>
to QVector<QgsLineString*>. Add it to python binding.

lbartoletti committed:
- Add a convenient method for checking validity of the geometry
- Apply change requested for qgsexpression

lbartoletti committed:
- Write conversion method surfaceToPolygon with tests
- Fix setExteriorRing and one test

lbartoletti committed:
Reimplementation of moveVertex's method.

lbartoletti committed:
Add two constructors (QgsPoint, QPointF)

lbartoletti committed:
Adapt python binding

lbartoletti committed:
Resolve conflicts for qgsexpression

lbartoletti committed:
Try to resolve conflicts

lbartoletti committed:
Change method for perpendicularSegment to avoid pointer.

lbartoletti committed:
Merge branch 'master' of https://github.com/QGIS/QGIS into qgstriangle

lbartoletti committed:
Merge branch 'master' into qgstriangle

lbartoletti committed:
Reindent

nyalldawson commented:
@lbartoletti where's this at now? Ready to go?

@rouault can you re-review following all the updates?

lbartoletti commented:
It's ok for me.
Thank you.

nyalldawson commented:
Fantastic work - thank you!
